File: bootstrap/astnode.cpp
#include "astnode.h"
#include "errors.h"

#include <cassert>
#include <cctype>
#include <sstream>

using namespace simplex;

static bool isWhitespace(char c) {
  return strchr(" \t\n\r", c) != nullptr;
}

static void expect(NodeKind kind, ASTInput& input, const std::string& token) {
  size_t tokenSize = token.size();
  if (tokenSize > input.size()) {
    const auto remaining = input.remaining();
    throw ParseError(kind, token.c_str(), remaining.c_str());
  }
  const std::string shouldBeToken(input.get(), tokenSize);
  if (shouldBeToken != token) {
    throw ParseError(kind, token.c_str(), shouldBeToken.c_str());
  }
  input.advance(tokenSize);
}

ASTNode::ASTNode(NodeKind kind) : m_kind(kind) { }

void ASTNode::toString(std::stringstream& ss) const {
  static int indentLevel = 0;
  assert(indentLevel >= 0);
  for (size_t i=0; i<indentLevel; i++) {
    ss << " ";
  }
  ss << NodeKindName(m_kind);
  ss << std::endl;
  indentLevel++;
  for (const auto& child : m_children) {
    child.toString(ss);
  }
  indentLevel--;
}

std::string ASTNode::toString() const {
  std::stringstream ss;
  this->toString(ss);
  return ss.str();
}

std::ostream& simplex::operator<<(std::ostream& stream, const ASTNode& node) {
  std::string result = node.toString();
  stream << result;
  return stream;
}

ASTNode ASTNode::parseProgram(const char *inputStream, size_t len) {
  ASTInput input(inputStream, len);
  return parseProgram(input);
}

ASTNode ASTNode::parseProgram(ASTInput& input) {
  auto ret = ASTNode(NodeKind::program);
  ret.m_children.push_back(parseExpression(input));
  if (input.size() > 0) {
    ret.m_children.push_back(parseProgram(input));
  }
  return ret;
}

ASTNode ASTNode::parseExpression(ASTInput& input) {
  NodeKind kind = NodeKind::expression;
  ASTNode ret(kind);
  parseOptionalWhitespace(input);
  if (input.size() == 0) {
    throw ParseError(kind, "(", "EOF");
  }
  char next = input.peek();
  if (next == '(') {
    expect(kind, input, "(");
    ret.m_children.push_back(parseExpression(input));
    ret.m_children.push_back(parseOptionalParameterList(input));
    parseOptionalWhitespace(input);
    expect(kind, input, ")");
  } else if (next == '\'' || std::isdigit(next)) {
    ret = parseLiteral(input);
  } else {
    ret = parseIdentifier(input);
  }
  parseOptionalWhitespace(input);
  return ret;
}

ASTNode ASTNode::parseLiteral(ASTInput& input) {
  NodeKind kind = NodeKind::literal;
  ASTNode ret(kind);
  if (input.size() == 0) {
    throw ParseError(kind, "any valid literal", "EOF");
  }
  if (input.peek() == '\'') {
    // string
    ret.m_children.push_back(parseString(input));
  } else {
    ret.m_children.push_back(parseNumber(input));
  }
  return ret;
}

ASTNode ASTNode::parseNumber(ASTInput& input) {
  NodeKind kind = NodeKind::number;
  
  // read digits until whitespace or decimal
  std::stringstream ss;
  bool isFloat = false;
  size_t inputLen = input.size();
  for (size_t i=0; i<inputLen; i++) {
    char next = input.peek();
    if (i > 0 && next == '.') {
      ss << next;
      isFloat = true;
    } else {
      if (isWhitespace(next) || next == ')') {
        // number is done
        break;
      }
      if (!std::isdigit(next)) {
        throw ParseError(kind, "digits 0 through 9", next);
      }
      ss << next;
    }
    input.next();
  }

  // broke out early or, hit EOF?
  // maybe we have a valid number at this point
  auto result = ss.str();
  if (isFloat) {
    kind = NodeKind::floatingPoint;
    ASTNode ret(kind);
    ret.m_float = std::atof(result.c_str());
    return ret;
  } else {
    kind = NodeKind::integer;
    ASTNode ret(kind);
    ret.m_int = std::atol(result.c_str());
    return ret;
  }
}

ASTNode ASTNode::parseString(ASTInput& input) {
  NodeKind kind = NodeKind::string;
  ASTNode ret(kind);
  std::stringstream ss;
  bool foundEndOfString = false;
  expect(kind, input, "'");
  while (input.size() != 0) {
    char next = input.peek();
    if (next == '\'') {
      foundEndOfString = true;
      break;
    }
    if (next == '\\') {
      // escape char
      if (input.size() < 2) {
        throw ParseError(kind, "any character followed by escape sequence (\\)", "EOF");
      }
      ss << input.next();
      next = input.peek();
    }

    ss << next;
    input.next();
  }
  if (!foundEndOfString) {
    throw ParseError(kind, "end of string marker (')", "EOF");
  }

  expect(kind, input, "'");
  ret.m_string = ss.str();
  return ret;
}

ASTNode ASTNode::parseIdentifier(ASTInput& input) {
  NodeKind kind = NodeKind::identifier;
  ASTNode ret(kind);
  if (input.size() == 0) {
    throw ParseError(kind, "any valid identifier", "EOF");
  }
  char next = input.peek();
  if (next == '\'') {
    throw ParseError(kind, "non-whitespace character other than '\\'', '(' and ')'", next);
  }
  std::stringstream ss;
  while (input.size() != 0) {
    next = input.peek();
    if (isWhitespace(next) ||
        next == ')') {
      break;
    }
    if (next == '(') {
      throw ParseError(kind, "non-whitespace character other than '('", next);
    }
    ss << next;
    input.next();
  }
  ret.m_string = ss.str();
  assert(ret.m_string.size() != 0);
  return ret;
}

void ASTNode::parseOptionalWhitespace(ASTInput& input) {
  if (input.size() == 0) {
    return;
  }
  if (!isWhitespace(input.peek())) {
    return;
  }
  parseWhitespace(input);
}

void ASTNode::parseWhitespace(ASTInput& input) {
  bool foundWhitespace = false;
  while (input.size() != 0) {
    char next = input.peek();
    if (isWhitespace(next)) {
      foundWhitespace = true;
    } else if (!foundWhitespace) {
      throw ParseError(NodeKind::whitespace, "Any of: ' ', \\r, \\n, \\t", next);
    } else {
      break;
    }
    input.next();
  }
}

ASTNode ASTNode::parseOptionalParameterList(ASTInput& input) {
  NodeKind kind = NodeKind::optionalParameterList;
  ASTNode ret(kind);
  if (input.peek() != ')') {
    ASTNode parameterList(NodeKind::parameterList);
    parameterList.parseParameterList(input);
    ret.m_children.push_back(parameterList);
  }
  return ret;
}

void ASTNode::parseParameterList(ASTInput& input) {
  m_children.push_back(parseExpression(input));
  parseOptionalWhitespace(input);
  if (input.size() == 0) {
    throw ParseError(NodeKind::parameterList, "end of expression ')'", "EOF");
  }
  if (input.peek() == ')') {
    // hit end of parameter list
    return;
  }
  // more parameters left to parse
  parseParameterList(input);
}

bool ASTNode::operator==(const ASTNode& other) const {
  if (m_kind != other.m_kind) {
    return false;
  }
  switch (m_kind) {
    case NodeKind::integer:
      return m_int == other.m_int;
    case NodeKind::floatingPoint:
      return m_float == other.m_float;
    case NodeKind::identifier:
    case NodeKind::string:
      return m_string == other.m_string;
    default:
      if (m_children != other.m_children) {
        return false;
      }
  }
  return true;
}

NodeKind ASTNode::kind() const {
  return m_kind;
}

const std::vector<ASTNode> ASTNode::children() const {
  return m_children;
}

double ASTNode::floatingPoint() const {
  assert(m_kind == NodeKind::floatingPoint);
  return m_float;
}

int64_t ASTNode::integer() const {
  assert(m_kind == NodeKind::integer);
  return m_int;
}

const std::string& ASTNode::string() const {
  assert(m_kind == NodeKind::string ||
         m_kind == NodeKind::identifier);
  return m_string;
}

ASTNode::ASTNode() : m_kind(NodeKind::invalid) { }

File: src/astnode.rs
use std::fmt::Display;

use crate::astinput::ASTInput;
use crate::errors::ParseError;

#[derive(Clone, Debug, PartialEq)]
enum ASTValue {
    Int(i64),
    Double(f64),
    String(String),
    Children(Vec<ASTNode>),
    Invalid,
}

#[derive(Clone, Debug, PartialEq)]
pub struct ASTNode {
    kind: NodeKind,
    value: ASTValue,

    line: i64,
    col: i64,
}

impl ASTNode {
    pub fn parse_program(input: &mut ASTInput) -> Result<ASTNode, ParseError> {
        let kind = NodeKind::Program;
        let line = input.line();
        let col = input.col();
        let mut children = Vec::new();
        match ASTNode::parse_expression(input) {
            Ok(v) => children.push(v),
            Err(e) => return Err(e),
        }
        if input.size() > 0 {
            match ASTNode::parse_program(input) {
                Ok(v) => children.push(v),
                Err(e) => return Err(e),
            }
        }
        Ok(ASTNode {
            kind,
            value: ASTValue::Children(children),
            line,
            col,
        })
    }

    pub fn parse_expression(input: &mut ASTInput) -> Result<ASTNode, ParseError> {
        let kind = NodeKind::Expression;

        match ASTNode::parse_optional_whitespace(input) {
            Ok(_) => (),
            Err(e) => return Err(e),
        }

        if input.size() == 0 {
            return Err(ParseError::from(
                kind,
                "(",
                "EOF",
                input.line(),
                input.col(),
            ));
        }

        let next = input.peek();
        let ret = if next == '(' {
            let mut children = Vec::new();
            match ASTNode::expect(kind, input, "(") {
                Ok(_) => (),
                Err(e) => return Err(e),
            }
            match ASTNode::parse_expression(input) {
                Ok(v) => children.push(v),
                Err(e) => return Err(e),
            }
            match ASTNode::parse_optional_parameter_list(input) {
                Ok(v) => children.push(v),
                Err(e) => return Err(e),
            }
            match ASTNode::parse_optional_whitespace(input) {
                Ok(_) => (),
                Err(e) => return Err(e),
            }
            match ASTNode::expect(kind, input, ")") {
                Ok(_) => (),
                Err(e) => return Err(e),
            }
            Ok(ASTNode {
                kind,
                value: ASTValue::Children(children),
                line: input.line(),
                col: input.col(),
            })
        } else if next == '\'' || next.is_digit(10) {
            ASTNode::parse_literal(input)
        } else {
            ASTNode::parse_identifier(input)
        };
        match ASTNode::parse_optional_whitespace(input) {
            Ok(_) => (),
            Err(e) => return Err(e),
        }
        ret
    }
    pub fn parse_optional_parameter_list(input: &mut ASTInput) -> Result<ASTNode, ParseError> {
        let kind = NodeKind::OptionalParameterList;
        let value = if input.peek() != ')' {
            let mut parameter_list = ASTNode {
                kind: NodeKind::ParameterList,
                value: ASTValue::Children(Vec::new()),
                line: input.line(),
                col: input.col(),
            };
            match parameter_list.parse_parameter_list(input) {
                Ok(_) => (),
                Err(e) => return Err(e),
            }
            vec![parameter_list]
        } else {
            Vec::new()
        };
        Ok(ASTNode {
            kind,
            value: ASTValue::Children(value),
            line: input.line(),
            col: input.col(),
        })
    }
    pub fn parse_parameter_list(&mut self, input: &mut ASTInput) -> Result<(), ParseError> {
        let mut new_children = Vec::new();
        match ASTNode::parse_expression(input) {
            Ok(v) => new_children.push(v),
            Err(e) => return Err(e),
        }
        match &mut self.value {
            ASTValue::Children(children) => children.extend(new_children),
            _ => panic!(),
        }

        match ASTNode::parse_optional_whitespace(input) {
            Ok(_) => (),
            Err(e) => return Err(e),
        }

        if input.size() == 0 {
            return Err(ParseError::from(
                NodeKind::ParameterList,
                "end of expression ')'",
                "EOF",
                input.line(),
                input.col(),
            ));
        }
        if input.peek() == ')' {
            // hit end of parameter list
            return Ok(());
        }
        // more parameters left to parse
        self.parse_parameter_list(input)
    }
    pub fn parse_literal(input: &mut ASTInput) -> Result<ASTNode, ParseError> {
        let kind = NodeKind::Literal;
        let line = input.line();
        let col = input.col();
        if input.size() == 0 {
            return Err(ParseError::from(
                kind,
                "any valid literal",
                "EOF",
                input.line(),
                input.col(),
            ));
        }
        let result = if input.peek() == '\'' {
            // string
            ASTNode::parse_string(input)
        } else {
            ASTNode::parse_number(input)
        };
        let value = match result {
            Err(e) => return Err(e),
            _ => vec![result.unwrap()],
        };
        Ok(ASTNode {
            kind,
            value: ASTValue::Children(value),
            line,
            col,
        })
    }
    pub fn parse_number(input: &mut ASTInput) -> Result<ASTNode, ParseError> {
        let mut kind = NodeKind::Integer;

        // read digits until whitespace or decimal
        let mut ss = String::new();
        let mut is_float = false;
        let input_len = input.size();
        for i in 0..input_len {
            let next = input.peek();
            if i > 0 && next == '.' {
                ss.push(next);
                is_float = true;
            } else {
                if is_whitespace(next) || next == ')' {
                    // number is done
                    break;
                }
                if !next.is_digit(10) {
                    return Err(ParseError::from(
                        kind,
                        "digits 0 through 9",
                        next.to_string(),
                        input.line(),
                        input.col(),
                    ));
                }
                ss.push(next);
            }
            input.next();
        }

        // broke out early or, hit EOF?
        // maybe we have a valid number at this point
        let result = ss;
        let line = input.line();
        let col = input.col();
        if is_float {
            kind = NodeKind::FloatingPoint;
            match result.parse::<f64>() {
                Ok(value) => Ok(ASTNode {
                    kind,
                    value: ASTValue::Double(value),
                    line,
                    col,
                }),
                Err(_e) => panic!(),
            }
        } else {
            kind = NodeKind::Integer;
            match result.parse::<i64>() {
                Ok(value) => Ok(ASTNode {
                    kind,
                    value: ASTValue::Int(value),
                    line,
                    col,
                }),
                Err(_e) => panic!(),
            }
        }
    }
    pub fn parse_optional_whitespace(input: &mut ASTInput) -> Result<(), ParseError> {
        if input.size() == 0 {
            return Ok(());
        }
        if !is_whitespace(input.peek()) {
            return Ok(());
        }
        ASTNode::parse_whitespace(input)
    }
    pub fn parse_whitespace(input: &mut ASTInput) -> Result<(), ParseError> {
        let mut found_whitespace = false;
        while input.size() != 0 {
            let next = input.peek();
            if is_whitespace(next) {
                found_whitespace = true;
            } else if !found_whitespace {
                return Err(ParseError::from(
                    NodeKind::Whitespace,
                    "Any of: ' ', \\r, \\n, \\t",
                    next.to_string(),
                    input.line(),
                    input.col(),
                ));
            } else {
                break;
            }
            input.next();
        }
        Ok(())
    }
    pub fn parse_identifier(input: &mut ASTInput) -> Result<ASTNode, ParseError> {
        let kind = NodeKind::Identifier;
        let line = input.line();
        let col = input.col();
        if input.size() == 0 {
            return Err(ParseError::from(
                kind,
                "any valid identifier",
                "EOF",
                input.line(),
                input.col(),
            ));
        }
        let mut next = input.peek();
        if next == '\'' {
            return Err(ParseError::from(
                kind,
                "non-whitespace character other than '\\'', '(' and ')'",
                next.to_string(),
                input.line(),
                input.col(),
            ));
        }
        let mut ss = String::new();
        while input.size() != 0 {
            next = input.peek();
            if is_whitespace(next) || next == ')' {
                break;
            }
            if next == '(' {
                return Err(ParseError::from(
                    kind,
                    "non-whitespace character other than '('",
                    next.to_string(),
                    line,
                    col,
                ));
            }
            ss.push(next);
            input.next();
        }
        assert!(ss.chars().count() != 0);
        Ok(ASTNode {
            kind,
            value: ASTValue::String(ss),
            line,
            col,
        })
    }

    pub fn parse_string(input: &mut ASTInput) -> Result<ASTNode, ParseError> {
        let kind = NodeKind::String;
        let line = input.line();
        let col = input.col();
        let mut ss = String::new();
        let mut found_end_of_string = false;
        match ASTNode::expect(kind, input, "'") {
            Ok(_) => (),
            Err(e) => return Err(e),
        }
        while input.size() != 0 {
            let mut next = input.peek();
            if next == '\'' {
                found_end_of_string = true;
                break;
            }
            if next == '\\' {
                // escape char
                if input.size() < 2 {
                    return Err(ParseError::from(
                        kind,
                        "any character followed by escape sequence (\\)",
                        "EOF",
                        input.line(),
                        input.col(),
                    ));
                }
                ss.push(input.next());
                next = input.peek();
            }

            ss.push(next);
            input.next();
        }
        if !found_end_of_string {
            return Err(ParseError::from(
                kind,
                "end of string marker (')",
                "EOF",
                input.line(),
                input.col(),
            ));
        }

        match ASTNode::expect(kind, input, "'") {
            Ok(_) => (),
            Err(e) => return Err(e),
        }
        Ok(ASTNode {
            kind,
            value: ASTValue::String(ss),
            line,
            col,
        })
    }

    // produces invalid node! should only use for testing
    pub fn new() -> ASTNode {
        ASTNode {
            kind: NodeKind::Invalid,
            value: ASTValue::Invalid,
            line: 0,
            col: 0,
        }
    }

    pub fn kind(&self) -> NodeKind {
        self.kind
    }

    pub fn children(&self) -> &Vec<ASTNode> {
        match &self.value {
            ASTValue::Children(children) => children,
            _ => panic!(),
        }
    }

    pub fn integer(&self) -> i64 {
        match self.value {
            ASTValue::Int(i) => i,
            _ => panic!(),
        }
    }

    pub fn floating_point(&self) -> f64 {
        match self.value {
            ASTValue::Double(d) => d,
            _ => panic!(),
        }
    }

    pub fn string(&self) -> String {
        match &self.value {
            ASTValue::String(s) => s.clone(),
            _ => panic!(),
        }
    }

    pub fn line(&self) -> i64 {
        self.line
    }

    pub fn col(&self) -> i64 {
        self.col
    }

    fn expect<S: AsRef<str>>(
        kind: NodeKind,
        input: &mut ASTInput,
        token: S,
    ) -> Result<(), ParseError> {
        let token_str = token.as_ref();
        let token_size = token_str.chars().count();
        if token_size > input.size() {
            let remaining = input.get();
            return Err(ParseError::from(
                kind,
                token,
                remaining,
                input.line(),
                input.col(),
            ));
        }
        let should_be_token = &input.get()[0..token_size];
        if should_be_token != token_str {
            return Err(ParseError::from(
                kind,
                token,
                should_be_token,
                input.line(),
                input.col(),
            ));
        }
        input.advance(token_size);
        Ok(())
    }
}

impl Display for ASTNode {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        let mut ss = String::new();
        // unsafe because of static mutable INDENT_LEVEL
        unsafe {
            assert!(INDENT_LEVEL >= 0);
            for _i in 0..INDENT_LEVEL {
                ss.push(' ');
            }
            let kind = format!("{:#?}", self.kind());
            ss.push_str(&kind);
            ss.push('\n');
            INDENT_LEVEL += 1;
            for child in self.children() {
                let child_str = child.to_string();
                ss.push_str(&child_str);
            }
            INDENT_LEVEL -= 1;
        }
        write!(f, "{}", ss)
    }
}

#[derive(Clone, Copy, Debug, PartialEq)]
pub enum NodeKind {
    Expression,
    FloatingPoint,
    Identifier,
    Integer,
    Invalid,
    Literal,
    OptionalParameterList,
    ParameterList,
    Program,
    String,
    Whitespace,
}

fn is_whitespace(c: char) -> bool {
    c == ' ' || c == '\t' || c == '\n' || c == '\r'
}

static mut INDENT_LEVEL: i64 = 0;

#[cfg(test)]
mod tests {
    // Note this useful idiom: importing names from outer (for mod tests) scope.
    use super::*;

    fn whitespace_variations(input: Vec<String>) -> Vec<String> {
        // inserts whitespace before, after, and before/after
        let mut ret = Vec::new();
        for str in input {
            ret.push(str.clone());
            ret.push(format!("  {}", str.clone()));
            ret.push(format!("{}  ", str.clone()));
            ret.push(format!(" {} ", str));
        }
        ret
    }
    fn identifiers() -> Vec<String> {
        vec!["identifier", "foo", "@#*&%&$#", "...", "💩"]
            .iter()
            .map(|x| x.to_string())
            .collect()
    }

    fn strings() -> Vec<String> {
        vec!["'foo bar'", "'&\"\\'+~💩$'"]
            .iter()
            .map(|x| x.to_string())
            .collect()
    }

    fn integers() -> Vec<String> {
        vec!["0", "1", "928453821"]
            .iter()
            .map(|x| x.to_string())
            .collect()
    }

    fn floats() -> Vec<String> {
        vec!["0.23592", "29384."]
            .iter()
            .map(|x| x.to_string())
            .collect()
    }

    fn literals() -> Vec<String> {
        let mut ret = Vec::new();
        ret.extend(strings());
        ret.extend(integers());
        ret.extend(floats());
        ret
    }

    fn expressions() -> Vec<String> {
        let mut ret: Vec<String> = vec!["( + 3 4)", " (  +  3  4  ) ", "(- 1.5)", "(* (- 1.5) 2)"]
            .iter()
            .map(|x| x.to_string())
            .collect();
        ret.extend(whitespace_variations(identifiers()));
        ret.extend(whitespace_variations(literals()));
        ret
    }

    fn programs() -> Vec<String> {
        let mut ret = expressions();
        for i in 0..expressions().len() - 1 {
            let item1 = ret[i].clone();
            let item2 = ret[i + 1].clone();
            let combined = item1 + " " + &item2;
            ret.push(combined);
        }
        ret
    }

    #[test]
    fn test_parse_program() {
        for str in programs() {
            let mut input = ASTInput::from_str(&str);
            let result = ASTNode::parse_program(&mut input);
            match result {
                Ok(node) => assert_eq!(node.kind(), NodeKind::Program),
                Err(_) => assert!(false),
            }
        }
    }

    #[test]
    fn test_parse_expression() {
        for str in expressions() {
            let mut input = ASTInput::from_str(&str);
            let result = ASTNode::parse_expression(&mut input);
            match result {
                // resulting NodeKind could be any of:
                // * expression
                // * literal
                // * identifier
                // due to simplified parse tree structure (useless expressions are skipped).
                Ok(node) => assert!(
                    node.kind == NodeKind::Expression
                        || node.kind == NodeKind::Literal
                        || node.kind == NodeKind::Identifier
                ),
                Err(_) => assert!(false),
            }
        }
    }

    #[test]
    fn test_parse_identifier() {
        for str in identifiers() {
            let mut input = ASTInput::from_str(&str);
            let result = ASTNode::parse_identifier(&mut input);
            match result {
                Ok(node) => assert_eq!(node.kind(), NodeKind::Identifier),
                Err(_) => assert!(false),
            }
        }
    }

    #[test]
    fn test_parse_literal() {
        for str in literals() {
            let mut input = ASTInput::from_str(&str);
            let result = ASTNode::parse_literal(&mut input);
            match result {
                Ok(node) => assert_eq!(node.kind(), NodeKind::Literal),
                Err(_) => assert!(false),
            }
        }
    }

    #[test]
    fn test_parse_string() {
        for str in strings() {
            let mut input = ASTInput::from_str(&str);
            let result = ASTNode::parse_string(&mut input);
            match result {
                Ok(node) => assert_eq!(node.kind(), NodeKind::String),
                Err(_) => assert!(false),
            }
        }
    }

    #[test]
    fn test_parse_number() {
        for str in floats() {
            let mut input = ASTInput::from_str(&str);
            let result = ASTNode::parse_number(&mut input);
            match result {
                Ok(node) => assert_eq!(node.kind(), NodeKind::FloatingPoint),
                Err(_) => assert!(false),
            }
        }
        for str in integers() {
            let mut input = ASTInput::from_str(&str);
            let result = ASTNode::parse_number(&mut input);
            match result {
                Ok(node) => assert_eq!(node.kind(), NodeKind::Integer),
                Err(_) => assert!(false),
            }
        }
    }
}
